cluster_management/src/main/java/com/pinterest/rocksplicator/Participant.java

```diff
@@ -444,4 +369,209 @@ public void run() {
+     * 4. Wait until all partition replicas hosted by this participant to have moved to OFFLINE state.
+     *
+     * 5. Now we are free to disconnect to HelixManager, so that all listeners and message callback
+     * handlers are properly un-registered.
+     */
+    if (helixManager != null && helixManager.isConnected()) {
+      disableInstance(helixManager.getClusterManagmentTool(), staticClusterName, staticInstanceId);
+    }
+
+    if (helixManager != null) {
+      new HelixManagerShutdownHook(helixManager).run();
+      helixManager = null;
+    }
+    if (staticParticipantLeaderEventsLogger != null) {
+      LOG.error("Stopping Participant LeaderEventsLogger");
+      staticParticipantLeaderEventsLogger.close();
+      staticParticipantLeaderEventsLogger = null;
+    }
+  }
+
+  private static void registerInstanceDomain(
+      final HelixAdmin helixAdmin,
+      final String clusterName,
+      final String instanceName,
+      final String domainName) {
+    HelixConfigScope scope =
+        new HelixConfigScopeBuilder(HelixConfigScope.ConfigScopeProperty.PARTICIPANT)
+            .forCluster(clusterName).forParticipant(instanceName)
+            .build();
+    Map<String, String> properties = new HashMap<String, String>();
+    properties.put("DOMAIN", domainName + ",instance=" + instanceName);
+    helixAdmin.setConfig(scope, properties);
+  }
+
+  private static void disableInstance(
+      final HelixAdmin helixAdmin,
+      final String clusterName,
+      final String instanceId) {
+    retryWithBoundedExponentialBackOff(
+        new ExceptionalAction() {
+          @Override
+          public void apply() throws Exception {
+            LOG.error("Disabling the instance to force offlining existing resources");
+            helixAdmin.enableInstance(clusterName, instanceId, false);
+          }
+        },
+        new ExceptionalCondition() {
+          @Override
+          public boolean apply() throws Exception {
+            // When the instance is no longer enabled
+            return !helixAdmin.getInstanceConfig(clusterName, instanceId).getInstanceEnabled();
+          }
+        }, 10);
+
+    retryWithBoundedExponentialBackOff(
+        new ExceptionalAction() {
+          @Override
+          public void apply() throws Exception {
+            LOG.error(
+                "Adding \"disabled\" instance tag, to force spectator to remove this instance from "
+                    + "shard_map ");
+            helixAdmin.addInstanceTag(clusterName, instanceId, "disabled");
+          }
+        },
+        new ExceptionalCondition() {
+          @Override
+          public boolean apply() throws Exception {
+            // When the instance config tag contains the "disabled" tag.
+            return helixAdmin.getInstanceConfig(clusterName, instanceId).containsTag("disabled");
+          }
+        }, 10);
+
+    /**
+     * Here we need a mechanism to wait for all instances to go offline. However currently that logic
```

Comment:
Why is it complicated? Can we look at the current state for the instance and check if everything is offline?

Reply:
Exactly.. that is too much of the code to do it along with this PR. I want to add that and other function like maintenance tag... so that if the node is to be taken down for scheduled maintenance or rotation... we don't allow participant to successfully deploy / restart on such instances..

Reply:
Ok. I tried to get ZNRecord for the instance, but can't get anything out of it. There isn't any direct api available that provides the information needed to ensure that all partitions are offlined. 

One way is to iterate through all resources and their corresponding externalViews to make sure the instance doesn't have any non-offlined partition. However that approach is too risky, for one that externalViews are bulky in nature and getting all externalViews in a large cluster during deploy could potentially increase the zk load more then is necessary.

So, for now, we will wait it out for 5 seconds which seems sufficient time to get the data offlined.

In the same way, we also don't have any way to see if all partitions are fully operational. The reason is that when an instance comes up it may not have any assigned partition or controller may have rebalanced them to other participants. Hence there is no straight way to say that all partitions have moved from offline state to next.

We might want to consider enabling a delay once a statemodel has been registered with helixManager to give it a time
to online (or move them to follower, bootstrap or slave state transition) partitions for pre-existing data partitions.

Reply:
I have tested on dev cluster to make sure this change does disable the instance and offline existing top state partitions when going down. While coming back up, it does wait to enable the instance and remove the disabled flag.

It also make sure that if "maintenance" instance tag has been applied to an instance, it won't allow the Participant to successfully online.